File: frontend/lib/pdf/builder.ts
import { COMPANY_INFO } from "@/config/company.config";
import jsPDF from "jspdf";

// ── Shared color palette ─────────────────────────────────
export const COLORS = {
  purple: [124, 58, 237] as [number, number, number],
  darkText: [26, 26, 26] as [number, number, number],
  mutedText: [120, 120, 120] as [number, number, number],
  lightBorder: [220, 220, 220] as [number, number, number],
  rowAlt: [252, 252, 252] as [number, number, number],
  headerFill: [245, 245, 245] as [number, number, number],
  totalFill: [237, 233, 254] as [number, number, number],
};

// ── Shared layout constants ──────────────────────────────
export const PAGE = {
  width: 210,
  height: 297,
  marginLeft: 18,
  get marginRight() {
    return this.width - this.marginLeft;
  },
};

// ── Formatting helpers ───────────────────────────────────
export function fmtCurrency(n: number): string {
  return new Intl.NumberFormat("en-NG", {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(n);
}

export function fmtDate(iso: string | null | undefined): string {
  if (!iso) return "—";
  return new Date(iso).toLocaleDateString("en-GB", {
    day: "2-digit",
    month: "short",
    year: "numeric",
  });
}

export function cap(s: string): string {
  if (!s) return s;
  return s.charAt(0).toUpperCase() + s.slice(1);
}

// ── Image loading ─────────────────────────────────────────
export async function loadImageAsBase64(
  url: string
): Promise<{ dataUrl: string; width: number; height: number } | null> {
  try {
    const res = await fetch(url);
    if (!res.ok) return null;
    const blob = await res.blob();
    const dataUrl = await new Promise<string>((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = () => resolve(reader.result as string);
      reader.onerror = reject;
      reader.readAsDataURL(blob);
    });
    const dims = await new Promise<{ width: number; height: number }>((resolve) => {
      const img = new Image();
      img.onload = () => resolve({ width: img.naturalWidth, height: img.naturalHeight });
      img.onerror = () => resolve({ width: 1, height: 1 });
      img.src = dataUrl;
    });
    return { dataUrl, ...dims };
  } catch {
    return null;
  }
}

// ── Document header (logo + company info) ───────────────
export async function drawHeader(doc: jsPDF): Promise<void> {
  const { marginLeft: ml, marginRight: mr } = PAGE;
  const logoDataUrl = await loadImageAsBase64(COMPANY_INFO.logoPath);

  if (logoDataUrl) {
    const logoH = 15;
    const logoW = (logoDataUrl.width / logoDataUrl.height) * logoH;
    doc.addImage(logoDataUrl.dataUrl, "PNG", ml, 10, logoW, logoH);
  } else {
    doc.setFontSize(16);
    doc.setFont("helvetica", "bold");
    doc.setTextColor(...COLORS.darkText);
    doc.text(COMPANY_INFO.name, ml, 20);
  }

  doc.setFontSize(7.5);
  doc.setFont("helvetica", "normal");
  doc.setTextColor(...COLORS.mutedText);
  doc.text(COMPANY_INFO.tagline, ml, 28);

  doc.setFontSize(7.5);
  doc.setTextColor(...COLORS.mutedText);
  doc.text(COMPANY_INFO.address, mr, 14, { align: "right" });
  doc.text(`Tel: ${COMPANY_INFO.phone}  |  ${COMPANY_INFO.email}`, mr, 19, { align: "right" });
  doc.text(COMPANY_INFO.website, mr, 24, { align: "right" });

  drawDivider(doc, 33);
}

// ── Footer ────────────────────────────────────────────────
export function drawFooter(doc: jsPDF, disclaimer: string, pageLabel = "Page 1"): void {
  const { width: pageW, height: pageH } = PAGE;
  doc.setFontSize(7);
  doc.setFont("helvetica", "normal");
  doc.setTextColor(...COLORS.mutedText);
  doc.text(disclaimer, pageW / 2, pageH - 10, { align: "center" });
  doc.text(pageLabel, pageW / 2, pageH - 6, { align: "center" });
}

// ── Divider line ──────────────────────────────────────────
export function drawDivider(doc: jsPDF, y: number): void {
  const { marginLeft: ml, marginRight: mr } = PAGE;
  doc.setDrawColor(...COLORS.lightBorder);
  doc.line(ml, y, mr, y);
}

// ── Document title block ─────────────────────────────────
export function drawTitle(
  doc: jsPDF,
  title: string,
  subtitle: string,
  titleY = 44,
  subtitleY = 51,
  dividerY = 56
): void {
  const { marginLeft: ml } = PAGE;

  doc.setFontSize(22);
  doc.setFont("helvetica", "bold");
  doc.setTextColor(...COLORS.purple);
  doc.text(title, ml, titleY);

  doc.setFontSize(8.5);
  doc.setFont("helvetica", "normal");
  doc.setTextColor(...COLORS.mutedText);
  doc.text(subtitle, ml, subtitleY);

  drawDivider(doc, dividerY);
}

// ── Label/value pair (two-column info grid) ──────────────
export function drawLabelValue(
  doc: jsPDF,
  label: string,
  value: string,
  x: number,
  y: number,
  maxWidth: number
): void {
  doc.setFontSize(6.5);
  doc.setFont("helvetica", "bold");
  doc.setTextColor(...COLORS.mutedText);
  doc.text(label.toUpperCase(), x, y);

  doc.setFontSize(9.5);
  doc.setFont("helvetica", "bold");
  doc.setTextColor(...COLORS.darkText);
  const lines = doc.splitTextToSize(value || "—", maxWidth);
  doc.text(lines[0] ?? "—", x, y + 5);
}

// ── Generic table renderer ───────────────────────────────
export interface TableColumn<T> {
  header: string;
  width: number; // mm
  align: "left" | "center" | "right";
  render: (row: T) => string;
}

export function drawTable<T>(
  doc: jsPDF,
  columns: TableColumn<T>[],
  rows: T[],
  startY: number
): number {
  const { marginLeft: ml, marginRight: mr } = PAGE;
  const tableW = mr - ml;
  const hdrH = 9;
  const rowH = 8;

  const cx: number[] = [ml];
  columns.slice(0, -1).forEach((col, i) => cx.push(cx[i] + col.width));

  let y = startY;

  // Header
  doc.setFillColor(...COLORS.headerFill);
  doc.rect(ml, y, tableW, hdrH, "F");
  doc.setDrawColor(...COLORS.lightBorder);
  doc.rect(ml, y, tableW, hdrH, "S");

  doc.setFontSize(7.5);
  doc.setFont("helvetica", "bold");
  doc.setTextColor(90, 90, 90);

  columns.forEach((col, i) => {
    const tx =
      col.align === "right" ? cx[i] + col.width - 2 :
      col.align === "center" ? cx[i] + col.width / 2 :
      cx[i] + 2;
    doc.text(col.header, tx, y + 6, { align: col.align });
  });

  y += hdrH;

  // Rows
  rows.forEach((row, idx) => {
    if (idx % 2 === 1) {
      doc.setFillColor(...COLORS.rowAlt);
      doc.rect(ml, y, tableW, rowH, "F");
    }
    doc.setDrawColor(235, 235, 235);
    doc.line(ml, y + rowH, ml + tableW, y + rowH);

    doc.setFontSize(9);
    doc.setFont("helvetica", "normal");
    doc.setTextColor(...COLORS.darkText);

    columns.forEach((col, i) => {
      const text = col.render(row);
      const tx =
        col.align === "right" ? cx[i] + col.width - 2 :
        col.align === "center" ? cx[i] + col.width / 2 :
        cx[i] + 2;

      if (col.align === "left") {
        const lines = doc.splitTextToSize(text, col.width - 4);
        doc.text(lines[0] ?? "", tx, y + 5.5);
      } else {
        doc.text(text, tx, y + 5.5, { align: col.align });
      }
    });

    y += rowH;
  });

  return y; // returns the y position after the table, for continued layout
}

// ── Total / summary highlight row ────────────────────────
export function drawTotalRow(
  doc: jsPDF,
  label: string,
  value: string,
  y: number,
  labelWidth: number
): number {
  const { marginLeft: ml, marginRight: mr } = PAGE;
  const tableW = mr - ml;
  const rowH = 8;

  doc.setFillColor(...COLORS.totalFill);
  doc.rect(ml, y, tableW, rowH + 2, "F");

  doc.setFontSize(9);
  doc.setFont("helvetica", "bold");
  doc.setTextColor(...COLORS.purple);
  doc.text(label, ml + 2, y + 6.5);
  doc.text(value, mr - 2, y + 6.5, { align: "right" });

  return y + rowH + 2;
}

File: backend/app/shared/pdf/builder.py
from __future__ import annotations

from io import BytesIO
from urllib.request import urlopen

from reportlab.lib import colors
from reportlab.lib.pagesizes import A4
from reportlab.lib.units import mm
from reportlab.lib.utils import ImageReader
from reportlab.pdfbase.pdfmetrics import stringWidth
from dataclasses import dataclass
from typing import Callable, Generic, Sequence, TypeVar



from datetime import date, datetime
from decimal import Decimal

from app.shared.config.company import COMPANY_INFO


# ---------------------------------------------------------------------
# Page
# ---------------------------------------------------------------------

PAGE_WIDTH, PAGE_HEIGHT = A4

MARGIN_LEFT = 18 * mm
MARGIN_RIGHT = PAGE_WIDTH - (18 * mm)


# ---------------------------------------------------------------------
# Colors
# ---------------------------------------------------------------------

PURPLE = colors.HexColor("#7C3AED")

DARK_TEXT = colors.HexColor("#1A1A1A")

MUTED_TEXT = colors.HexColor("#787878")

LIGHT_BORDER = colors.HexColor("#DCDCDC")

ROW_ALT = colors.HexColor("#FCFCFC")

HEADER_FILL = colors.HexColor("#F5F5F5")

TOTAL_FILL = colors.HexColor("#EDE9FE")


T = TypeVar("T")


@dataclass(frozen=True)
class TableColumn(Generic[T]):
    header: str
    width: float
    align: str = "left"          # left | center | right
    renderer: Callable[[T], str] = str


def fmt_currency(value: Decimal | float | int) -> str:
    return f"{float(value):,.2f}"


def fmt_date(value: date | datetime | None) -> str:
    if not value:
        return "—"

    if isinstance(value, datetime):
        value = value.date()

    return value.strftime("%d %b %Y")


def capitalize(value: str | None) -> str:
    if not value:
        return ""

    return value[:1].upper() + value[1:]

_cached_logo: ImageReader | None = None


def get_logo() -> ImageReader | None:
    global _cached_logo

    if _cached_logo is None:
        _cached_logo = load_logo()

    return _cached_logo


def load_logo() -> ImageReader | None:
    """
    Download the company logo from Cloudinary.

    Returns None if unavailable so callers can gracefully
    fall back to rendering the company name.
    """

    try:
        response = urlopen(COMPANY_INFO.logo_url)

        return ImageReader(
            BytesIO(
                response.read(),
            )
        )

    except Exception:
        return None
    
class PdfBuilder:

    def __init__(self, canvas):
        self.canvas = canvas

        self.page_width = PAGE_WIDTH
        self.page_height = PAGE_HEIGHT

        self.margin_left = MARGIN_LEFT
        self.margin_right = MARGIN_RIGHT

        self.cursor_y = PAGE_HEIGHT
    
    def draw_divider(
        self,
        y: float,
    ) -> None:

        self.canvas.setStrokeColor(LIGHT_BORDER)

        self.canvas.line(
            self.margin_left,
            y,
            self.margin_right,
            y,
        )
    def draw_header(self) -> None:

        logo = get_logo()

        if logo:
            self.canvas.drawImage(
                logo,
                self.margin_left,
                PAGE_HEIGHT - 25 * mm,
                width=28 * mm,
                height=15 * mm,
                preserveAspectRatio=True,
                mask="auto",
            )

        else:
            self.canvas.setFont(
                "Helvetica-Bold",
                16,
            )

            self.canvas.setFillColor(DARK_TEXT)

            self.canvas.drawString(
                self.margin_left,
                PAGE_HEIGHT - 18 * mm,
                COMPANY_INFO.name,
            )

        #
        # Company information
        #

        self.canvas.setFillColor(MUTED_TEXT)

        self.canvas.setFont(
            "Helvetica",
            7.5,
        )

        self.canvas.drawString(
            self.margin_left,
            PAGE_HEIGHT - 30 * mm,
            COMPANY_INFO.tagline,
        )

        right_x = self.margin_right

        self.canvas.drawRightString(
            right_x,
            PAGE_HEIGHT - 16 * mm,
            COMPANY_INFO.address,
        )

        self.canvas.drawRightString(
            right_x,
            PAGE_HEIGHT - 21 * mm,
            f"Tel: {COMPANY_INFO.phone} | {COMPANY_INFO.email}",
        )

        self.canvas.drawRightString(
            right_x,
            PAGE_HEIGHT - 26 * mm,
            COMPANY_INFO.website,
        )
        self.draw_divider(
            PAGE_HEIGHT - 33 * mm,
        )
        self.cursor_y = PAGE_HEIGHT - 42 * mm
    

    def draw_title(
        self,
        title: str,
        subtitle: str,
    ) -> None:
        """
        Draw the document title and subtitle.
        """

        #
        # Leave some space below the company header
        #
        y = self.cursor_y - 3 * mm

        #
        # Title
        #
        self.canvas.setFont(
            "Helvetica-Bold",
            22,
        )
        self.canvas.setFillColor(PURPLE)

        self.canvas.drawString(
            self.margin_left,
            y,
            title,
        )

        #
        # Subtitle
        #
        self.canvas.setFont(
            "Helvetica",
            8.5,
        )
        self.canvas.setFillColor(MUTED_TEXT)

        self.canvas.drawString(
            self.margin_left,
            y - 8 * mm,
            subtitle,
        )

        #
        # Divider
        #
        divider_y = y - 13 * mm

        self.draw_divider(divider_y)

        #
        # Leave space before the next section
        #
        self.cursor_y = divider_y - 8 * mm

    def draw_label_value(
        self,
        label: str,
        value: str,
        x: float,
        y: float,
        width: float,
    ) -> float:
        """
        Draw a label with a bold value beneath it.

        The value is automatically wrapped to fit within the
        supplied width.

        Returns the vertical space consumed.
        """

        #
        # Label
        #
        self.canvas.setFont(
            "Helvetica-Bold",
            6.5,
        )

        self.canvas.setFillColor(MUTED_TEXT)

        self.canvas.drawString(
            x,
            y,
            label.upper(),
        )

        #
        # Value
        #
        text = value or "—"

        font_name = "Helvetica-Bold"
        font_size = 9.5

        self.canvas.setFont(
            font_name,
            font_size,
        )

        self.canvas.setFillColor(DARK_TEXT)

        #
        # Wrap text
        #
        words = text.split()

        lines: list[str] = []
        current = ""

        for word in words:

            candidate = (
                word
                if not current
                else f"{current} {word}"
            )

            if (
                stringWidth(
                    candidate,
                    font_name,
                    font_size,
                )
                <= width
            ):
                current = candidate

            else:
                if current:
                    lines.append(current)

                current = word

        if current:
            lines.append(current)

        #
        # Draw wrapped lines
        #
        line_spacing = 4 * mm
        line_y = y - 5 * mm

        for line in lines:

            self.canvas.drawString(
                x,
                line_y,
                line,
            )

            line_y -= line_spacing

        #
        # Return height consumed so callers can
        # lay out the next section correctly.
        #
        return (
            5 * mm +
            max(1, len(lines)) * line_spacing
        )

    def draw_footer(
        self,
        disclaimer: str,
        page_number: int = 1,
    ) -> None:

        self.canvas.setFont(
            "Helvetica",
            7,
        )

        self.canvas.setFillColor(MUTED_TEXT)

        self.canvas.drawCentredString(
            PAGE_WIDTH / 2,
            10 * mm,
            disclaimer,
        )

        self.canvas.drawCentredString(
            PAGE_WIDTH / 2,
            6 * mm,
            f"Page {page_number}",
        )
    
    def next_line(self, height: float) -> float:
        """
        Move the cursor down and return the new Y position.
        """

        self.cursor_y -= height
        return self.cursor_y
    
    def draw_table(
        self,
        columns: Sequence[TableColumn],
        rows: Sequence,
        row_height: float = 8 * mm,
    ) -> None:
        """
        Draw a reusable table.

        Left-aligned cells automatically wrap onto multiple lines,
        while centered and right-aligned cells remain single-line.
        """

        table_width = sum(col.width for col in columns)

        x_positions = [self.margin_left]

        for column in columns[:-1]:
            x_positions.append(
                x_positions[-1] + column.width
            )

        header_height = 9 * mm

        y = self.cursor_y

        #
        # Header background
        #
        self.canvas.setFillColor(HEADER_FILL)

        self.canvas.rect(
            self.margin_left,
            y - header_height,
            table_width,
            header_height,
            fill=1,
            stroke=0,
        )

        #
        # Header border
        #
        self.canvas.setStrokeColor(LIGHT_BORDER)

        self.canvas.rect(
            self.margin_left,
            y - header_height,
            table_width,
            header_height,
            fill=0,
            stroke=1,
        )

        #
        # Header text
        #
        self.canvas.setFont(
            "Helvetica-Bold",
            7.5,
        )

        self.canvas.setFillColor(MUTED_TEXT)

        for i, column in enumerate(columns):

            x = x_positions[i]

            if column.align == "right":

                self.canvas.drawRightString(
                    x + column.width - 2 * mm,
                    y - 6 * mm,
                    column.header,
                )

            elif column.align == "center":

                self.canvas.drawCentredString(
                    x + column.width / 2,
                    y - 6 * mm,
                    column.header,
                )

            else:

                self.canvas.drawString(
                    x + 2 * mm,
                    y - 6 * mm,
                    column.header,
                )

        y -= header_height

        #
        # Body
        #
        for index, row in enumerate(rows):

            #
            # Determine row height based on wrapped cells
            #
            max_lines = 1

            for column in columns:

                if column.align != "left":
                    continue

                text = str(column.renderer(row))

                words = text.split()

                current = ""
                lines = 0

                for word in words:

                    candidate = (
                        word
                        if not current
                        else f"{current} {word}"
                    )

                    if (
                        stringWidth(
                            candidate,
                            "Helvetica",
                            9,
                        )
                        <= column.width - 4 * mm
                    ):
                        current = candidate

                    else:
                        lines += 1
                        current = word

                if current:
                    lines += 1

                max_lines = max(
                    max_lines,
                    lines,
                )

            actual_row_height = max(
                row_height,
                max_lines * 4 * mm + 4 * mm,
            )

            #
            # Alternate background
            #
            if index % 2 == 1:

                self.canvas.setFillColor(ROW_ALT)

                self.canvas.rect(
                    self.margin_left,
                    y - actual_row_height,
                    table_width,
                    actual_row_height,
                    fill=1,
                    stroke=0,
                )

            #
            # Bottom border
            #
            self.canvas.setStrokeColor(
                colors.HexColor("#EBEBEB")
            )

            self.canvas.line(
                self.margin_left,
                y - actual_row_height,
                self.margin_left + table_width,
                y - actual_row_height,
            )

            self.canvas.setFont(
                "Helvetica",
                9,
            )

            self.canvas.setFillColor(DARK_TEXT)

            #
            # Draw cells
            #
            for i, column in enumerate(columns):

                text = str(
                    column.renderer(row)
                )

                x = x_positions[i]

                if column.align == "right":

                    self.canvas.drawRightString(
                        x + column.width - 2 * mm,
                        y - 5.5 * mm,
                        text,
                    )

                elif column.align == "center":

                    self.canvas.drawCentredString(
                        x + column.width / 2,
                        y - 5.5 * mm,
                        text,
                    )

                else:

                    #
                    # Wrap text
                    #
                    words = text.split()

                    current = ""

                    lines: list[str] = []

                    for word in words:

                        candidate = (
                            word
                            if not current
                            else f"{current} {word}"
                        )

                        if (
                            stringWidth(
                                candidate,
                                "Helvetica",
                                9,
                            )
                            <= column.width - 4 * mm
                        ):
                            current = candidate

                        else:
                            if current:
                                lines.append(current)
                            current = word

                    if current:
                        lines.append(current)

                    line_y = y - 5.5 * mm

                    for line in lines:

                        self.canvas.drawString(
                            x + 2 * mm,
                            line_y,
                            line,
                        )

                        line_y -= 4 * mm

            y -= actual_row_height

        self.cursor_y = y

    def draw_total_row(
        self,
        label: str,
        value: str,
    ) -> None:
        """
        Purple highlighted total row.
        """

        row_height = 10 * mm

        width = self.margin_right - self.margin_left

        self.canvas.setFillColor(TOTAL_FILL)

        self.canvas.rect(
            self.margin_left,
            self.cursor_y - row_height,
            width,
            row_height,
            fill=1,
            stroke=0,
        )

        self.canvas.setFont(
            "Helvetica-Bold",
            9,
        )

        self.canvas.setFillColor(PURPLE)

        self.canvas.drawString(
            self.margin_left + 2 * mm,
            self.cursor_y - 6 * mm,
            label,
        )

        self.canvas.drawRightString(
            self.margin_right - 2 * mm,
            self.cursor_y - 6 * mm,
            value,
        )

        self.cursor_y -= row_height + 2 * mm


    def draw_text(
        self,
        text: str,
        font_size: float = 9,
        color=DARK_TEXT,
    ) -> None:
        """
        Draw a simple paragraph at the current cursor.
        """

        self.canvas.setFont(
            "Helvetica",
            font_size,
        )

        self.canvas.setFillColor(color)

        self.canvas.drawString(
            self.margin_left,
            self.cursor_y,
            text,
        )

        self.cursor_y -= 6 * mm
        

        



        

    



